### DISCUSSION TITLE
Install packages using doom-emacs(straight.el)

### DISCUSSION THREAD

**Discussion by hwiorn** [kjambunathan/org-mode-ox-odt#133] [Show and tell]
Write below in `~/.doom.d/packages.el`, then call `~/.emacs.d/bin/doom sync`

```emacs
(package! JabrefExportChicagoODF
  :recipe (:host github :repo "kjambunathan/JabRefChicagoForOrgmode"))
(package! ox-odt
  :recipe (:host github :repo "kjambunathan/org-mode-ox-odt"
           :files ("lisp/ox-odt.el"
                   "lisp/contrib/ox-jabref.el"
                   "etc"
                   "docs"
                   "contrib/odt/LibreOffice")))
```

Doom-emacs uses `straight.el`, so it may be like this in the `straight.el` way. 

```emacs
(use-package JabrefExportChicagoODF
  :straight (:host github :repo "kjambunathan/JabRefChicagoForOrgmode")
(use-package ox-odt
  :straight (org-mode-ox-odt :host github :repo "kjambunathan/org-mode-ox-odt"
                             :files ("lisp/ox-odt.el"
                                     "lisp/contrib/ox-jabref.el"
                                     "etc"
                                     "docs"
                                     "contrib/odt/LibreOffice")))
```

**Comment by ghost** [comment]
I'm trying the second version and it doesn't seem to work. The first use-package snippet tells me that use-package can't load JabRefChicagoForOrgmode and the second one tells me the following:

```
$ cd /home/anon/.emacs.d/straight/repos/org-mode-ox-odt/
$ git submodule update --init --recursive

fatal: No url found for submodule path 'contrib/odt/JabRefChicagoForOrgmode' in .gitmodules

[Return code: 128]
```

**Reply by QiangF** [reply]
JabRef is not needed any more. See https://github.com/kjambunathan/org-mode-ox-odt/issues/164

**Reply by dlukes** [reply]
Which means you can add `:nonrecursive t` to the `:recipe` definition, which will avoid fetching JabRef. I'm currently using the following to load the forked `ox-odt` on Doom Emacs and it works fine:

```elisp
(package! ox-odt
  :recipe (:host github :repo "kjambunathan/org-mode-ox-odt" :nonrecursive t
           :files ("lisp/ox-odt.el"
                   "etc"
                   "docs"
                   "contrib/odt/LibreOffice")))
```

**Comment by kjambunathan** [comment]
Update your `package-archives` variable, as seen in https://kjambunathan.gitlab.io/org-mode-ox-odt/Getting-Started-with-ODT-export.html#Installation

Then do `M-x package-refresh-contents` and `M-x package-install ox-odt`.  That should get the job done.  It would help if you also do `M-x package-install citeproc`.  (`citeproc` is not a big deal, if you aren't interested in using new citation features)

-----

Another alternative would be to download the `ox-odt` tar from https://github.com/kjambunathan/kjambunathan.github.io/tree/master/elpa.  And then do `M-x package-install-file <ox-odt-<whatever>.tar`.

----- 

I don't use straight, my guess is this would work i.e., Just skip the jabref parts. 

```elisp
(use-package ox-odt
  :straight (org-mode-ox-odt :host github :repo "kjambunathan/org-mode-ox-odt"
                             :files ("lisp/ox-odt.el"
                                     "etc"
                                     "docs"
                                     "contrib/odt/LibreOffice")))
```
 Org has a new citation syntax, and and the CSL-based citation engine is much much superior to what JabRef could provide.

----

It would help if you install `org-9.5.*` from GNU ELPA.

**Comment by kjambunathan** [comment]
This is probably the most fool proof way, and this is what I do myself.

Do a git checkout of this repo, and assuming that you have checked out the source under `~/src1` subdir, add the following at the very bottom of your init file ... and restart Emacs.

```elisp
(setq load-path (cons "~/src1/org-mode-ox-odt/lisp/" load-path))
(load-library "ox-odt")

```

**Comment by hwiorn** [comment]
I've not used this package for a while because I've been really busy at work and home.

For the record, it needed a step more for latest doomemacs before using `(use-package ox-odt)`.
Latest Emacs and `package!` macro doesn't override `ox-odt` feature, So `unload-feautre` or `load-library` is needed before using it for overriding.

`straight.el` handles `load-path` itself, it needs to change the path order like @kjambunathan 's [snippet](https://github.com/kjambunathan/org-mode-ox-odt/discussions/133#discussioncomment-2214682)

```
(after! doom-packages
  (setq load-path (cons
                   (doom-path straight-base-dir "straight" straight-build-dir "ox-odt")
                   load-path))
  (load-library "ox-odt"))

(use-package! ox-odt
  :ensure t
  :config
  (setq org-latex-to-mathml-convert-command "latex2mathml -f %I > %o")) ;; latex2mathml
```

**Comment by Tonus1** [comment]
Hi all
I struggle a lot trying to install this authoritative fork.
I've tryed almost all what I read on the help and issues with no luck.
Please note that I am using Doom Emacs and I am a beginner with (Doom) Emacs.
```
generated  févr. 20, 2024 00:50:05
system     Slackware 15.0 x86_64 Linux 5.15.145 x86_64
emacs      27.2 ~/.config/emacs/
doom       3.0.0-pre PROFILE=_@0 grafted, HEAD -> master, origin/master, origin/HEAD 9863985 2023-10-07 02:33:45 +0200 ~/.config/doom
shell      /bin/bash
features   XPM JPEG TIFF GIF PNG RSVG SOUND GPM DBUS GLIB NOTIFY INOTIFY ACL GNUTLS LIBXML2 FREETYPE HARFBUZZ XFT ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS JSON UNEXEC LCMS2 GMP
traits     batch server-running envvar-file custom-file
modules    :config use-package :completion company vertico :ui doom doom-dashboard hl-todo modeline ophints (popup +defaults) (vc-gutter +pretty) vi-tilde-fringe workspaces :editor (evil +everywhere) file-templates fold snippets
	   :emacs (dired +icons) electric (ibuffer +icons) undo vc :term eshell vterm :checkers syntax (spell +flyspell) :tools (eval +overlay) lookup magit pdf :os tty :lang emacs-lisp json latex lua markdown (org +journal
	   +roam2 +pretty +publish) php python sh yaml :app calendar emms irc :config (default +bindings +smartparens)
packages   (flycheck-aspell) (dmenu) (emojify) (evil-tutor) (imenu-list) (ivy-posframe) (org-auto-tangle) (org-web-tools) (ox-gemini) (password-store) (rainbow-mode) (resize-window) (beacon) (olivetti) (neotree) (sudo-edit)
	   (org-superstar) (org-trello) (imenu-list) (all-the-icons-ivy-rich) (counsel) (swiper)
```
Since I use stable version of Slackware I might have to go with a manual compile of a newer version. Let me know.

Regards

**Comment by hwiorn** [comment]
@Tonus1 

Try this.

#### package.el

```elisp
(package! ox-odt
  :recipe
  (:host github
   :repo "kjambunathan/org-mode-ox-odt"
   :files ("lisp/ox-odt.el"
           "lisp/ox-ods.el"
           "lisp/odt.el"
           "etc"
           "docs"
           "contrib/odt/LibreOffice")))
```
#### config.el
```elisp
;; https://github.com/kjambunathan/org-mode-ox-odt
(with-eval-after-load 'doom-packages
  (setq load-path (cons
                   (doom-path straight-base-dir "straight" straight-build-dir "ox-odt")
                   load-path)))

(use-package! ox-odt
  :config
  (require 'ox-ods)
  ;; (setq  org-latex-to-mathml-convert-command "texmath -f tex -t mathml %I > %o") ;; textmath
  (setq org-latex-to-mathml-convert-command "latex2mathml -f %I > %o") ;; latex2mathml
  )

```

**Reply by Tonus1** [reply]
Thank you @hwiorn !

I tried and it seemed to instal the package.
But another probleme arised, as I was missing peg package wich I added.
I now have another error peg related 
```
Error caused by user's config or system: /home/antoine/.config/doom/config.el, (void-function byte-run--set-speed)
```
```
> Checking Doom Emacs...
  x There was an unexpected runtime error
    Message: Symbol's function definition is void
    Details: (byte-run--set-speed)
    Backtrace:
      (byte-run--set-speed id nil -1)
      (list (byte-run--set-speed id nil -1) (cons 'put (cons (list 'quote id) '('byte-optimizer #'byte-compile-inline-expand))))
      (progn (list (byte-run--set-speed id nil -1) (cons 'put (cons (list 'quote id) '('byte-optimizer #'byte-compile-inline-expand)))))
      (if inline (progn (list (byte-run--set-speed id nil -1) (cons 'put (cons (list 'quote id) '('byte-optimizer #'byte-compile-inline-expand))))))
      (cons (list 'put (list 'quote id) ''peg--rule-definition (list 'quote exp)) (if inline (progn (list (byte-run--set-speed id nil -1) (cons 'put (cons (list 'quote id) '('byte-optimizer #'byte-compile-inline-expand)))))))
      (cons 'eval-and-compile (cons (list 'put (list 'quote id) ''peg--rule-definition (list 'quote exp)) (if inline (progn (list (byte-run--set-speed id nil -1) (cons 'put (cons (list 'quote id) '('byte-optimizer #'byte-com...
      (list 'progn (list 'defalias (list 'quote id) (list 'peg--lambda (list 'quote pexs) args (if inline (apply #'peg--translate exp) (peg--translate-rule-body name exp)))) (cons 'eval-and-compile (cons (list 'put (list 'qu...
      (let ((id (peg--rule-id name)) (exp (peg-normalize (cons 'and pexs)))) (list 'progn (list 'defalias (list 'quote id) (list 'peg--lambda (list 'quote pexs) args (if inline (apply #'peg--translate exp) (peg--translate-ru...
      (let ((inline nil)) (while (keywordp (car pexs)) (let* ((val (car-safe (prog1 pexs (setq pexs (cdr pexs)))))) (if (eq val :inline) (progn (setq inline (car pexs))) nil)) (setq pexs (cdr pexs))) (let ((id (peg--rule-id ...
      ((closure (t) (name args &rest pexs) "Define PEG rule NAME as equivalent to PEXS.\nThe PEG expressions in PEXS are implicitly combined with the\nsequencing `and' operator of PEG grammars." (let ((inline nil)) (while (k...
      (define-peg-rule null nil :inline t (guard t))
      (eval-buffer #<buffer  *load*-12074> nil "/home/antoine/.config/emacs/.local/straight/build-27.2/peg/peg.el" nil t)
```
Before and after the peg install the test to check if the new package is used failed

**Comment by hwiorn** [comment]
I'm using Emacs 29.2. So I can't give you a specific hint for resolving byte-compile error.

Try removing `peg` depedency(Remove the   `(require 'ox-ods)` part in both use-package! And package! And try install it again).
You need to remove ox-odt directory manually before `doom sync` again. lets see if it works out.

**Reply by hwiorn** [reply]
I mean, The "(require 'ox-odd)" part, not "(require 'peg)" in ox-ods.el directly.

**Comment by kjambunathan** [comment]
---

You may  need all of `peg`, `seq`, `map`  and `rnc-mode`.  (There are all "standard" Emacs packages, and you can pick them from GNU ELPA or GNU-devel ELPA.

`peg` ([1](https://elpa.gnu.org/packages/peg.html), [2](https://elpa.gnu.org/devel/peg.html)) is needed only for `ox-ods`, the `ODS spreadhseet Exporter`.

`seq` ([1](https://elpa.gnu.org/packages/seq.html), [2](https://elpa.gnu.org/devel/seq.html)), which is fairly an young library, is used by both ODT and ODS exporters

`map` ([1](https://elpa.gnu.org/packages/map.html), [2](https://elpa.gnu.org/devel/map.html)), which is fairly an young library, is used by both ODT and ODS exporters

Modify  the below settings to you taste.  Note that the numbers represent the priority of archives.  (Or you can probably install from `https://kjambunathan.github.io/elpa/` archive, which has a "old" ODT and ODS exporters.  The archive hasn't been updated in a while, but I have lots of house-keeping to do before I update my ELPA archive)

```lisp
(custom-set-variables
 '(package-archive-priorities
   '(("gnu" . 100) ("nongnu" . 90) ("gnu-devel" . 80) ("melpa" . 70)
     ("ox-odt" . 60)))
 '(package-archives
   '(("gnu-devel" . "https://elpa.gnu.org/devel/")
     ("gnu" . "https://elpa.gnu.org/packages/")
     ("nongnu" . "https://elpa.nongnu.org/nongnu/")
     ("melpa" . "https://melpa.org/packages/")
     ("ox-odt" . "https://kjambunathan.github.io/elpa/"))))
```

-------

Related issues

- [Cannot compile cfb100ebdbb1a913a412bb9e975e1e4: Error No such file or directory, peg](https://github.com/kjambunathan/org-mode-ox-odt/issues/264).  **Remark**: You can skip this issue if you are NOT compiling from source.  But there are some observations made there which could be "useful"
- [No `seq-keep` defined](https://github.com/kjambunathan/org-mode-ox-odt/issues/266#issuecomment-1896225249)

---

Number of users of this ODT exporter can be counted in one's 10 fingers.  So, the user count doesn't justify significant investment of my time in keeping house in order.  When I make a next release, I will keep all these complaints in mind.

I have more things planned for ODT and ODS exporter though ... So, thanks for your interest and being patient with some annoyances.

Some quick workarounds

- Remove `ox-ods.el` from from doom install spec (This will remove dependency on `ox-ods` and `peg`)
- Install from kjambunathan.github.io/elpa.  A old install which will likely go through
- Download `tar` from https://github.com/kjambunathan/kjambunathan.github.io/tree/master/elpa, and M-x package-install-file
- Pin to a "old" commit SHA and install from that
- Or do the whole dance and install peg, seq, map and rnc-mode.  If installing these from GNU ELPA doesn't work, install them from GNU-devel ELPA (see above snippet for URLs)

----

If you run in to issues, let me know. I will be happy to help

**Comment by kjambunathan** [comment]
Message: Symbol's function definition is void
    Details: (byte-run--set-speed)

This problem is occurring while compiling 'peg'. I run Emacs from git
checkout  ... and peg is also fairly new.

If you are NOT exporting to ODS files (excel spreadsheets),you will not
need 'ox-ods.el'.
You will need odt.el and ox-odt.el and style stuff in etc directory... That
is about it
Just by looking at the error it looks like a compatibility issue.

What is the version of peg you are using ... Make sure you are installing
peg from GNU ELPA and not MELPA

M-x find-library peg and see the version header at the top of the file

Also upadte this issue with emacs --version or start doing M-x
report-emacs-bug,the new mail buffer will tell you about your Emacs version
and whether or not it is "AOT" (ahead of time) compiled ...

This seems like more a compatibilty issue which will automatically
disappear after some time with no efforts on our side

Just copying @ouboub.  He is the one originally reported issues with peg
and seq libraries.  He is THE ONLY ONE USER who cares about ODS exporter.
Tomorrow I will try to track down when "byte-run--set-speed" was introduced
and what its speed is

For now just remove "ox-ods.el" from the below snippet and try again ....


(package! ox-odt

  :recipe

  (:host github

   :repo "kjambunathan/org-mode-ox-odt"

   :files ("lisp/ox-odt.el"

           "lisp/ox-ods.el"

           "lisp/odt.el"

           "etc"

           "docs"

           "contrib/odt/LibreOffice")))


You will get the ODT exporter .... and possibly some issues with seq and
map.  These issues with seq and map could be easily managed

**Comment by kjambunathan** [comment]
emacs      27.2



That is an OLD version.   I am not surprised you are having
incompatibilities issue.


Any way you can upgrade Emacs to recent stable release Emacs 29.2 was
released only months ago ...

Is there a way to pin a repo to specific git commit .... If then try a
version that 6 months old, 1 yr old.  You get the idea.  (Look at the
commit history on github and any "big gap" in commits means those
commit hashes arr worth pinning to

**Comment by kjambunathan** [comment]
Look at commit messages in kjambunathan.github.io ELPA repo
https://github.com/kjambunathan/kjambunathan.github.io/tree/master/elpa

The last release was made some 6 months ago and corresponded to commit

This release corresponds to commit 36840ddd6d345ba473a586ceec986aad9946b5a3

So try pinning straight.el's pull to above commit hash

I believe it will work

**Comment by Tonus1** [comment]
Thanks both for your time. I installed seq map and rnc-mode with no changes on the error.

I will try to upgrade emacs before anything else.

Regards
